node/lib/util/stitch_util.js

```diff
@@ -0,0 +1,637 @@
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+"use strict";
+
+const assert         = require("chai").assert;
+const co             = require("co");
+const NodeGit        = require("nodegit");
+
+const Commit              = require("./commit");
+const DoWorkQueue         = require("./do_work_queue");
+const GitUtil             = require("./git_util");
+const SubmoduleConfigUtil = require("./submodule_config_util");
+const SubmoduleUtil       = require("./submodule_util");
+const TreeUtil            = require("./tree_util");
+
+const FILEMODE            = NodeGit.TreeEntry.FILEMODE;
+
+/**
+ * Return a string having the value of the specified `sha` with the "/"
+ * character inserted between the second and third characters of `sha`.  The
+ * behavior is undefined unless `sha` is at least three characters long.
+ *
+ * @param {String} sha
+ * @return {String}
+ */
+exports.splitSha = function (sha) {
+    const pre = sha.slice(0, 2);
+    const post = sha.slice(2);
+    return `${pre}/${post}`;
+};
+
+/**
+ * Return the name of the ref used to indicate that the commit having the
+ * specified `sha` has beren converted.
+ *
+ * @param {String} sha
+ * @return {String}
+ */
+exports.convertedRefName = function (sha) {
+    return "refs/stitched/converted/" + exports.splitSha(sha);
+};
+
+/**
+ * Return the name of the ref indicating that the submodule commit having the
+ * specified `subSha` has been fetched for the specified `metaSha`; this ref
+ * also serves to root that submodule commit.
+ *
+ * @param {String} metaSha
+ * @param {String} subSha
+ * @return {String}
+ */
+exports.fetchedSubRefName = function (metaSha, subSha) {
+    return "refs/stitched/fetched/" + exports.splitSha(metaSha) + "/sub/" +
+        exports.splitSha(subSha);
+};
+
+/**
+ * Return a summary for the specified `commit` for the submodule having the
+ * specifiewd `name` to be used to describe changes that are rolled into a
+ * meta-repo commit.
+ *
+ * @param {String} name
+ * @param {Commit} subCommit
+ */
+exports.summarizeSubCommit = function (name, subCommit) {
+    assert.isString(name);
+    assert.instanceOf(subCommit, NodeGit.Commit);
+
+    const author = subCommit.author();
+    return `\
+Includes changes from submodule ${name} on ${subCommit.id().tostrS()}.
+Author: ${author.name()} <${author.email()}>
+Date:   ${Commit.formatCommitTime(author.when())}
+
+${subCommit.message()}
+`;
+};
+
+/**
+ * Return true if there is no difference in the specified `newTree` and
```

Comment:
It's confusing to have "isEmpty" be some sort of equality comparison thing.

Also, an empty tree is illegal, but maybe we use it for some special thing.

Reply:
Yes, that's misleading. It should be something like `isUnchanged`.